### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) for Inko

### DISCUSSION THREAD

**Discussion by zamazan4ik** [inko-lang/inko#769] [General]
Hi!

I noticed that in the `Cargo.toml` [file](https://github.com/inko-lang/inko/blob/main/Cargo.toml) Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48). If for some distros LTO creates problems (I've seen multiple issues with LTO in AUR e.g.) - maintainers can opt out of LTO in their specific cases.

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I have made quick tests (Apple M1, macOS): enabling `lto = true` reduced the binary size from 99 Mib to 91 Mib. The project also builds fine with LTO on my Linux setup (Fedora).

Thank you.

**Comment by yorickpeterse** [comment]
This is better discussed in an issue, so I [created one based on this discussion](https://github.com/inko-lang/inko/issues/770). Let's move things there.
